x-pack/plugins/security_solution/server/endpoint/services/actions/action_details_by_id.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core/server';
+import { getActionCompletionInfo, mapToNormalizedActionRequest } from './utils';
+import {
+  ActionDetails,
+  ActivityLogAction,
+  ActivityLogActionResponse,
+  EndpointAction,
+  EndpointActionResponse,
+  EndpointActivityLogAction,
+  EndpointActivityLogActionResponse,
+  LogsEndpointAction,
+  LogsEndpointActionResponse,
+} from '../../../../common/endpoint/types';
+import {
+  ACTION_REQUEST_INDICES,
+  ACTION_RESPONSE_INDICES,
+  catchAndWrapError,
+  categorizeActionResults,
+  categorizeResponseResults,
+  getUniqueLogData,
+} from '../../utils';
+import { EndpointError } from '../../../../common/endpoint/errors';
+import { NotFoundError } from '../../errors';
+import { ACTIONS_SEARCH_PAGE_SIZE } from './constants';
+
+export const getActionDetailsById = async (
+  esClient: ElasticsearchClient,
+  actionId: string
+): Promise<ActionDetails> => {
+  let actionRequestsLogEntries: Array<ActivityLogAction | EndpointActivityLogAction>;
+
+  let normalizedActionRequest: ReturnType<typeof mapToNormalizedActionRequest> | undefined;
+  let actionResponses: Array<ActivityLogActionResponse | EndpointActivityLogActionResponse>;
+
+  try {
+    // Get both the Action Request(s) and action Response(s)
+    const [actionRequestEsSearchResults, actionResponsesEsSearchResults] = await Promise.all([
+      // Get the action request(s)
+      esClient
+        .search<EndpointAction | LogsEndpointAction>(
+          {
+            index: ACTION_REQUEST_INDICES,
+            body: {
+              query: {
+                bool: {
+                  filter: [{ term: { action_id: actionId } }],
+                },
+              },
+            },
+          },
+          {
+            ignore: [404],
+          }
+        )
+        .catch(catchAndWrapError),
+
+      // Get the Action Response(s)
+      esClient
+        .search<EndpointActionResponse | LogsEndpointActionResponse>(
+          {
+            index: ACTION_RESPONSE_INDICES,
+            size: ACTIONS_SEARCH_PAGE_SIZE,
+            body: {
+              query: {
+                bool: {
+                  filter: [{ term: { action_id: actionId } }],
```

Comment:
Can we not run both queries at once, across both indices? This may be accomplishable in one roundtrip

Reply:
We're already doing that here with `Promise.all`.

Reply:
I don't mean send two ES requests simultaneously. I mean sending _a single_ ES request telling it to search 2 indices at once. Which, using only the new-action-data format, we may still be able to do. Will save a few layers of overhead

Reply:
Oh, I see. Yes, that would be ideal.

Reply:
@pzl I'll look into that. Q. Do you mean using the normal `.search()` method? or using the "bulk" (I think elastic has such a thing) api?

fyi - may push it out of this PR. I'll see

Reply:
normal search. You can input multiple indices I believe. And since the schemas for the fields you're filtering by are, I think, all in the same place. It's the same query.

Reply:
ahh. Ok. but that will make it bit more difficult to parse out the search result hits with the types and stuff. Let me look into it.

Reply:
Closing the loop on this one:

@pzl and I tried for a little bit to get this to work across the multiple indexes and different data shapes, but were not successful. We'll leave it as is for now (two concurrent requests).

Reply:
@paul-tavares I can help with this if you need it. Or we can do this in a new PR too.

Reply:
yeah, its probably not needed for this PR. We can defer it for another one if necessary